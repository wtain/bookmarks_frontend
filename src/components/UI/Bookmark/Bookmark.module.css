
.bookmark {
    border: black solid 1px;
    border-radius: 15px;
    background-color: darkgray;
    width: 100%-20px;
    margin: 5px;
    padding: 10px;
}

.id {
    font-weight: bold;
}

.summary {
    font-weight: bold;
    color: black;
    width: calc(100% - 30px);
}

.done {
    text-decoration: line-through;
}

.created {
    font-size: small;
}

.contents {
    white-space: pre-line;
    background: transparent;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    resize: none;
    width: 100%;
    color: black;
}

.contents:hover {
    border-style: inset;
}

.btn_remove {
    float: right;
    border-radius: 30px;
    background-color: darkgray;
    font-size: 20px;
}

.tags_container {
    width: calc(100% - 7px);    
}