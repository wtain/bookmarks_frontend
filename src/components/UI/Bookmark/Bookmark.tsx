
import React, { useEffect, useState } from "react";
import BookmarkDto from "../../../domain/dto/BookmarkDto";
import cl from './Bookmark.module.css'
import Collapsible from 'react-collapsible';
import ReactHelpers from "../../../utils/ReactHelpers";
import BookmarkEdit from "./detail/BookmarkEdit";
import BookmarkHeader from "./detail/BookmarkHeader";
import TagsEditor from "../TagsEditor/TagsEditor";
import TagDto from "../../../domain/dto/TagDto";
import TimeAgo from 'javascript-time-ago'

// import en from 'javascript-time-ago/locale/en.json'
import moment from "moment";


interface Props {
    bookmark: BookmarkDto
    doRemove: (bm: BookmarkDto) => void;
    doChangeContents: (new_contents: string) => void;
    doChangeSummary: (new_summary: string) => void;
    doAddTag: (tag: TagDto) => void;
    doRemoveTag: (index: number) => void;
    doChangeIsDone: (new_value: boolean) => void;
    highlightJustAdded?: boolean;
}

const Bookmark = (props: Props) => {

    const divRef = React.useRef<null | HTMLDivElement>(null);

    // const rtf = new Intl.RelativeTimeFormat('en', { style:'narrow'});
    // const elapsed = Date.now() - Date.parse(props.bookmark.created);
    // rtf.format(elapsed, "")

    const timeAgo = new TimeAgo('en-US');

    // hack
    const localFormat = "DD.MM.yyyy, HH:mm:ss";
    
    const createdAgo = timeAgo.format(moment(props.bookmark.created, localFormat).toDate());
    // const createdAgo = props.bookmark.created;

    let updatedAgo = undefined;
    if (props.bookmark.updated !== undefined) {
        updatedAgo = timeAgo.format(moment(props.bookmark.updated, localFormat).toDate());;
    }

    // useEffect(() => TimeAgo.addDefaultLocale(en));

    useEffect(() => {
        if (props.highlightJustAdded) {
            ReactHelpers.scrollToRef(divRef)
        }
    }, [props.highlightJustAdded])

    
    return (
        <div className={cl.bookmark} ref={divRef}>
            <Collapsible open={props.highlightJustAdded!} 
                         trigger={<BookmarkHeader bookmark={props.bookmark} 
                         doRemove={props.doRemove}
                         onIsDoneChanged={(new_value: boolean) => {
                             props.doChangeIsDone(new_value);
                         }}
                         onAcceptEdit={(new_summary: string) => {
                           props.doChangeSummary(new_summary);
                       }} />}>
                
                <div className={cl.created}>Created: {createdAgo}</div>
                {
                    (props.bookmark.updated !== undefined) ? 
                    <div className={cl.created}>Updated: {updatedAgo}</div> : <></>
                }

                <BookmarkEdit initialContents={props.bookmark.contents} 
                            onAcceptEdit={(new_contents: string) => {
                                props.doChangeContents(new_contents);
                            }} />
                <br />
                <TagsEditor 
                    containerClass={cl.tags_container}
                    tags={props.bookmark.tags} 
                    onTagAdded={(tag: TagDto) => {
                        props.doAddTag(tag);
                    }} 
                    onDelete={(index: number) => {
                        props.doRemoveTag(index);
                    }} />
            </Collapsible>
        </div>
    )
}

export default Bookmark;