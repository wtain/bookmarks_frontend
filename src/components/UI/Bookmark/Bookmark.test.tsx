import TimeAgo from 'javascript-time-ago';
import renderer from 'react-test-renderer';
import Bookmark from './Bookmark';
import en from 'javascript-time-ago/locale/en.json';
// import { render, screen } from '@testing-library/react';

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });


// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

TimeAgo.addDefaultLocale(en);
jest
  .useFakeTimers()
  .setSystemTime(new Date('2022-09-17T13:47:36.000Z'));

it('Bookmark renders correctly', () => {

  const bookmark = {
    summary: 'Edit tags',
    contents: 'Ability to edit tags for existing bookmarks\nSecond line\nThird line',
    id: 'ab2809a7-de53-4494-8ece-2aeecadce254',
    created: new Date("2022-01-30T20:48:19.000Z"),
    tags: [ { name: 'Task', id: '4b68bb47-c561-4958-8682-ea84ca39773d' } ],
    isDone: true,
    updated: new Date("2022-02-13T14:13:09.000Z")
  }

  const tree = renderer
    .create(
      <Bookmark bookmark={bookmark} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});