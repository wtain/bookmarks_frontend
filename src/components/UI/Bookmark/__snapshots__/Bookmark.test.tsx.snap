// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark renders correctly 1`] = `
<div
  className="bookmark"
>
  <input
    checked={true}
    className="done_checkbox"
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
  />
  <div
    className="header_div"
    data-for="registerTipab2809a7-de53-4494-8ece-2aeecadce254"
    data-tip={true}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span>
      <button
        className="btn_remove"
        onClick={[Function]}
      >
        ❌
      </button>
    </span>
    <div
      className="summary_preview"
    >
      <span
        className="summary_preview_text"
      />
      <a
        className="link"
        href="/bookmark/ab2809a7-de53-4494-8ece-2aeecadce254"
        onClick={[Function]}
      >
        🔗
      </a>
    </div>
    <div
      className="summary done"
      contentEditable={true}
      dangerouslySetInnerHTML={
        Object {
          "__html": "Edit tags",
        }
      }
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    />
  </div>
  <div
    className="created"
  >
    Created: 
    8 months ago
  </div>
  <div
    className="created"
  >
    Updated: 
    7 months ago
  </div>
  <div
    className="contents"
    contentEditable={true}
    dangerouslySetInnerHTML={
      Object {
        "__html": "Ability to edit tags for existing bookmarks
Second line
Third line",
      }
    }
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
  <br />
  <div
    className="container tags_container"
    onClick={[Function]}
  >
    <span
      className="tag"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Task
    </span>
    <span
      className="sp"
    >
      <input
        className="input"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Tags"
        value=""
      />
    </span>
  </div>
</div>
`;
